부산대_BE_진서현_5주차_1단계

jinseohyun1228 opened the pull request:
* step0에서는 이전 커밋 기록이 안뜨고 안전하게 PR을 날릴 수 있었는데 새로운 setp1에서는 다시 이전 커밋 기록들이 살아나서,,,
일단 step0으로 PR을 다시 날립니다! 
* 내일 낮에 해결방법이 있는지 확인하겠습니다.

## 구현기능
1. 카카오 로그인 구현

## 리뷰받고 싶은 부분
1. 현재 카카로 로그인 요청 Url은 kakao이고, 로그인을 처리하는 부분은 kakao/login으로 되어있습니다.    
이 uri이 RESTful API 설계에 적절한가요?

특히 wepConfig에서 허용 Url을 등록할때, meber 로그인 Url과 비슷하게 유지해야하나 라는 고민이 있습니다..!
```
    @Override
    public void addInterceptors(InterceptorRegistry registry) {
        // 인터셉터를 등록하는 메서드/members/register
        // CustomInterceptor를 등록하고, 모든 URL에 대해 인터셉터를 적용하도록 설정
        registry.addInterceptor(authLoginInterceptor)
                .excludePathPatterns("/members/login", "/members/join", "/kakao", "/", "/kakao","/kakao/login")
                .addPathPatterns("/**");
    }

```
2. 인가코드를 전달받아서 엑세스 토큰을 요청하고 유저를 등록하는 부분이 모두 하나의 메서드에 존재합니다.    
카카오 로그인이니 분리하기 어렵긴하지만 지금은 한 메서드가 몸집이 너무 큰 것 같습니다... 해결방법이 있을까요?


이번은 PR을 빨리 날리고 리뷰요청을 많이 드리고 싶어서 조금 급하게 코드를 짠 것 같습니다. 틈틈이 코드를 테스트하면서 추가로 궁금한것이 생기면 여쭤보겠습니다!

jinseohyun1228 committed:
git: 이전 코드 가져오기

jinseohyun1228 committed:
feat: 카카오 Oauth로그인에 사용되는 변수 프로퍼티 설정

jinseohyun1228 committed:
feat: 카카오 Oauth로그인에 사용되는 변수 프로퍼티 설정

jinseohyun1228 committed:
Merge branch 'step0' of https://github.com/jinseohyun1228/spring-gift-order into step0

jinseohyun1228 committed:
feat: 카카오 Oauth로그인에 사용되는 변수 프로퍼티 설정

jinseohyun1228 committed:
feat(KakaoTokenResponse) : 카카오 엑세스 토큰 응답 객체 생성

jinseohyun1228 committed:
feat: 카카오 유저정보 응답 객체 생성

jinseohyun1228 committed:
feat(KakaoController): 카카오 로그인 로직 컨트롤러 생성

<구현기능>
- KakaoController
1. requestKakaoLoginScreen() : 카카오 로그인 인증 코드를 요청하는 메서드
2. getKakaoCode : 전달받은 코드로 엑세스 토큰요청, 유저정보등록, jwt 발급하는 컨트롤러

jinseohyun1228 committed:
feat: 카카오 로그인 로직 서비스단 구현

<구현사항>
1. KakaoService 서비스 객체
(1) receiveKakaoToken() - 인가코드를 이용한 엑세스 토큰 발급해오기
(2) getKakaoProfile() - 엑세스 토큰을 이용한 유저정보 가져오기

2. UserService 서비스 객체
(1) loginKakaoUser() : 카카오 유저 정보를 받아서 없으면 DB에 저장하고, JWT를 발급

jinseohyun1228 committed:
feat(UserJoinRequest) : 카카오 유저 정보를 입력받아 생성하는 생성자 추가

jinseohyun1228 committed:
style: 코드 컨벤션

jinseohyun1228 committed:
style: 코드 컨벤션

jinseohyun1228 committed:
Merge branch 'jinseohyun1228' into step0

jinseohyun1228 committed:
fix: 카카오 로그인시에도 헤더에 jwt 담아서 요청

jinseohyun1228 committed:
git: 충돌 해결

hocaron commented on src/main/java/gift/main/config/KakaoProperties.java:
```diff
@@ -0,0 +1,74 @@
+@ConfigurationProperties(prefix = "spring.kakao")
+public class KakaoProperties {
+
+    private String grantType = "authorization_code";
```
해당 필드만 초기값을 둔 이유가 이유가 궁금합니다💭

jinseohyun1228 replied:
해당 부분은 구현하면서 필요성을 느껴 추가했습니다. 당시에는 문제를 느끼지 못했는데 지금은 통일성이 많이 떨어지네요... 

코드를 짠 후에 한번더 검토하는 과정이 필요할 것 같습니다 🥲 
캐치해주셔서 감사합니당 반영하겠습니다!

jinseohyun1228 replied:
반영했습니다!

hocaron commented on src/main/java/gift/main/config/KakaoProperties.java:
```diff
@@ -0,0 +1,74 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ConfigurationProperties(prefix = "spring.kakao")
+public class KakaoProperties {
```
setter, getter 없이 record 로 불변하게 구현해볼 수 있을 것 같아요😀

```
@ConfigurationProperties(prefix = "spring.kakao")
public record KakaoProperties(
        String grantType,
        String codeRequestUri,
        String tokenRequestUri,
        String userRequestUri,
        String redirectUri,
        String clientId,
        String password
) {
}
```

아래과 같이 `@ConfigurationPropertiesScan` 추가하면 빈등록이 됩니다😀
```java
@SpringBootApplication
@ConfigurationPropertiesScan
```

jinseohyun1228 replied:
이 부분 제가 잘 이해를 한 것인지 궁금합니다.

1. `@Configuration`는 애초에 필요없었음
2.  `@ConfigurationProperties`가 `application.properties`의 설정값을 주입해서 빈을 만들어줌, 그러나 주입시에 세터를 이용함
3. 클래스에서 레코드로 변경함에 따라 세터를 통해 주입이 불가능해져 `@ConfigurationPropertiesScan`을 사용함 이 어노테이션은 필드명을 기준으로 빈을 등록하기 때문에 레코드도 빈으로 등록할 수 있음

제가 잘 이해한 것이 맞을까요?!

hocaron replied:
@ConfigurationProperties 어노테이션은 그 자체로 빈 등록을 하지 않아요. 빈 등록을 위해서는 @ConfigurationProperties를 사용하는 클래스에 빈 등록을 할 수 있는 추가적인 어노테이션이나 메서드가 필요한데요, 서현님이 설정해주신 1번과 제가 제안드린 2번 방법이 있을 것 같아요.

1. @Configuration 추가
해당 클래스가 빈으로 등록

2. @EnableConfigurationProperties 추가
내부적으로 @ConfigurationProperties가 적용된 클래스를 스캔하여 빈으로 등록

저는 해당 클래스의 역할이 설정 클래스라기 보다는 속성 바인딩을 위한 클래스라고 생각해서 2번 방법을 제안 드렸습니다😀
record 사용시 불변객체로 사용할 수 있기도하고요. record 는 생성자를 통해 값을 설정하기 때문에 세터가 필요없습니다.

일반 클래스에서도 생성자를 통해 설정 파일의 값을 주입받을 수 있습니다. 이 경우에는 @ConstructorBinding 어노테이션을 사용해야합니다!

jinseohyun1228 replied:
헉 간단한 내용이라고만 생각햇는데 복잡하네요... 
더 알아보고 정리해보겠습니다

hocaron commented on src/main/java/gift/main/dto/KakaoTokenResponse.java:
```diff
@@ -0,0 +1,12 @@
+package gift.main.dto;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public record KakaoTokenResponse(
+        @JsonProperty("access_token") String accessToken,        //사용자 액세스 토큰 값
+        @JsonProperty("token_type") String tokenType,          //토큰타입, bearer로 고정
+        @JsonProperty("refresh_token") String refreshToken,       //사용자 리프레시 토큰 값
+        @JsonProperty("expires_in") int expiresIn,             //액세스 토큰과 ID 토큰의 만료 시간(초)
+        @JsonProperty("refresh_token_expires_in") int refreshTokenExpiresIn  //리프레시 토큰 만료 시간(초)
+) {
+}
```
모두 스네이크 케이스로 변환해야한다면 , JsonProperty를 각각 달아주는 것 대신 `@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)`를 사용해볼 수 있을 것 같아요.
```suggestion
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public record KakaoTokenResponse(
        String accessToken,        //사용자 액세스 토큰 값
        String tokenType,          //토큰타입, bearer로 고정
        String refreshToken,       //사용자 리프레시 토큰 값
        int expiresIn,             //액세스 토큰과 ID 토큰의 만료 시간(초)
        int refreshTokenExpiresIn  //리프레시 토큰 만료 시간(초)
) {
}
```

jinseohyun1228 replied:
헉 이 부분 진짜 몰랐습니다...!!
이런식으로 코드를 많이 짰었는데 처음알았습니다. 코드가 더 깔끔해지겠군요🔥🔥🔥 
반영하겠습니다 감사합니다.

jinseohyun1228 replied:
반영했습니다!

hocaron commented on src/main/java/gift/main/service/KakaoService.java:
```diff
@@ -0,0 +1,56 @@
+@Service
+public class KakaoService {
+
+    private static MediaType CONTENT_TYPE = MediaType.valueOf("application/x-www-form-urlencoded;charset=utf-8");
```
이미 정의된 값을 활용해서 사용해볼 수 있을 것 같아요.
```suggestion
private static final MediaType CONTENT_TYPE = new MediaType(APPLICATION_FORM_URLENCODED, UTF_8);
```

jinseohyun1228 replied:
헉 이런게 있었군요.. 뭔가 있을법한데 없다면 제가 못찾은 것이라고 생각하고 더 찾아보는 습관을 가져야할 것 같습니다...
덕분에 코드가 더 깔끔해졌습니다 감사합니다!

jinseohyun1228 replied:
반영했습니다 ^0^

hocaron commented:
1. 현재 카카로 로그인 요청 Url은 kakao이고, 로그인을 처리하는 부분은 kakao/login으로 되어있습니다.
이 uri이 RESTful API 설계에 적절한가요?

보통 로그인 처리하는 부분을 .../login/callback 으로 처리합니다!

2. 메서드 분리하면 좋을 것 같아요. 같이 리팩토링 해보시죠🚀

hocaron reviewed (approved):
서현님, 1단계 빠르게 구현해주셨네요👍

같이 작업하는 입장에서 궁금한 부분이 생겨 코멘트 드렸는데, 확인 부탁드릴게요!

jinseohyun1228 committed:
refactor(KakaoProperties): 레코드로 변경, grantType 설정값에서 가져오기, @@JsonNaming 사용

<변경사항>
1. KakaoController & KakaoService : 레코드 변경으로 getter메서드명 변경
2. Application : 레코드 변경으로 자동 빈 주입 방법 사용
3. application.properties : grantType값
4. KakaoProperties: 레코드로 변경, @Configuration삭제, @JsonNaming 추가

jinseohyun1228 committed:
test(KakaoPropertiesTest) : 카카오 프로퍼티 정상 빈 등록되었는지 검사

jinseohyun1228 committed:
--point : 혹시라도 컨트롤러 메서드 분리 안되면 여기로 와야함--

jinseohyun1228 committed:
refactor: KakaoProfile를 UserRequest처럼 다룰 수 있도록 변경

[변경 사항]
1. KakaoProfile : name, email, password, role 메서드 추가
2. User : KakaoProfile를 받아 생성하는 생성자 추가
3. UserJoinRequest : 필요 없어진 부분 삭제

[변경 후 ]
UserJoinRequest와 KakaoProfile간의 관계가 사라지고, KakaoProfile로만 login 과정을 처리할 수 있게 되어서 더 깔끔해짐

jinseohyun1228 committed:
refactor(KakaoController) : 코드-토큰-정보조회-가입-jwt이 한 메서드에 있던 것을 일단 가입과 jwt 발급 부분 분리

jinseohyun1228 committed:
refactor(UserService) : 카카오 로그인 과정 변경

jinseohyun1228 committed:
error : 카카오 컨트롤러 메서드 분리 문제

jinseohyun1228 commented:
### Rest Client 사용으로 카카오 컨트롤러 메서드 분리 어려움 

원래 흐름
-  인가코드를 전달받는 메서드에서  엑세스 토큰, 유저 정보, DB 저장, jwt 토큰을 다루고 있었습니다.

변경 후 
- 인가 코드를 전달 URL 쿼리파라미터가 존재하지만, 엑세스 토큰 URI는 쿼리 파라미터가 없다는 점에서 메서드를 분리함
- 따라서 인가코드를 전달받고, 해당 코드로 다시 요청을 날려 엑세스 토큰 URI 매핑 메서드에서 유 저 저장하려고 했다. 

그러나 Rest Client 사용으로 요청을 보낸후에 새로운 URL에서 처리하는게 아니라 이전 메서드에서 계속 처리하여 제 예상대로 되지 않습니다.

Rest CLient에 대해서 찾아봐도, 요청을 보내기만하는 로직은 찾아볼수가 없었는데요. 제가 못찾은건지 혹은 아예 없는 건지 궁금합니다.

이럴 경우에는 다른 라이브러리를 사용해야겠죠?

hocaron commented:
서현님 안녕하세요!

1. kakao/login 을 통해서 인가코드를 받아오고
2. kakao/login/callback 을 통해서, 1에서 받은 인가코드로 spring-gift 서비스의 토큰 생성

이렇게 2단계로 나누면 웹에서 2번 호출로 인증 로직을 모두 완료할 수 있을 것 같은데요😀처음에 아마 이렇게 구현하신 것 같은데, 현재 3단계로 분리된 이유가 있을까요?!

분리하면 해당 요청이 정말 kakao/login/callback(params="code")를 거쳐왔는지 확인하는 추가적인 원타임 토큰이 필요하지 않을까 싶어서 분리한 이유를 여쭤봅니다!

jinseohyun1228 commented:
>이렇게 2단계로 나누면 웹에서 2번 호출로 인증 로직을 모두 완료할 수 있을 것 같은데요😀처음에 아마 이렇게 구현하신 것 같은데, 현재 3단계로 분리된 이유가 있을까요?!

메서드를 분리해야하는 것이 2단계의 부분을 나눠야한다고 생각했습니다..!! 혹시 다른 부분이었을까요?!

jinseohyun1228 commented:
> 이럴 경우에는 다른 라이브러리를 사용해야겠죠?

아하!! ..!!! 넵!!

hocaron commented:
<img width="745" alt="image" src="https://github.com/user-attachments/assets/3661a953-f86c-4500-b8e0-e9bf19ba8ad4">

아하! 저는 공유해주신 그림에서 2, 3번이 합쳐지는게 api 호출면에서도 인증 로직이 분리되지 않아 응집성이 있다고 생각했는데요, 나누신 이유가 궁금해요!

jinseohyun1228 commented:
>아하! 저는 공유해주신 그림에서 2, 3번이 합쳐지는게 api 호출면에서도 인증 로직이 분리되지 않아 응집성이 있다고 생각했는데요, 나누신 이유가 궁금해요!

카카오컨트롤러는 모두 카카오 인증을 위한 역할을 가지고 있기 때문에 컨트롤러 내에서 메서드가 분리되어도 응집성에는 변화가 없을 것이라고 생각했습니다.

전체적으로는 2번 3번이 인증 로직에 포함되지만 코드발급과 토큰발급이라는 각각의 다른 역할을 수행하고 있어 세부 역할별로 메서드를 나누고 싶었습니다.

hocaron commented:
아하 네엡!

제가 그림을 이해를 못 한 것 같아요.

3번에서 말씀주신 jwt는 서비스를 이용하기 위해 발급한 토큰이라고 이해했는데, 2번에서 응답으로 받는 코드는 어떤 코드 일까요?

jinseohyun1228 commented:
>카카오컨트롤러는 모두 카카오 인증을 위한 역할을 가지고 있기 때문에 컨트롤러 내에서 메서드가 분리되어도 응집성에는 변화가 없을 것이라고 생각했습니다.
>
>전체적으로는 2번 3번이 인증 로직에 포함되지만 코드발급과 토큰발급이라는 각각의 다른 역할을 수행하고 있어 세부 역할별로 메서드를 나누고 싶었습니다.
>
>아하 네엡!
>
>제가 그림을 이해를 못 한 것 같아요.
>
>3번에서 말씀주신 jwt는 서비스를 이용하기 위해 발급한 토큰이라고 이해했는데, 2번에서 응답으로 받는 코드는 어떤 코드 일까요?


2번에서 받는 코드는 카카오 로그인화면에서 동의를 하고 카카오 서버에게  인가코드를 전달받습니다

3번은 사용자 유저 정보를 조회할 수 있는 엑세스 토큰을 전달받을 계획이었습니다

jinseohyun1228 committed:
study : 왜여기서 한지 모르겠지만 영속성 컨텍스트에 대한 테스트

jinseohyun1228 committed:
refactor: 카카오 인증 컨트롤러와 서비스단 구현

jinseohyun1228 committed:
fix : 카카오프로필에는 패스워드 값을 넣기 힘들어 중간 dto KakaoUser 추가

jinseohyun1228 committed:
fix : 설정값에 (")제거

jinseohyun1228 committed:
chore : 생성자 순서변경

jinseohyun1228 committed:
study : 여기서 왜 한건지 모르겠는 영속성 컨텍스트 테스트

jinseohyun1228 commented:
머지 부탁드립니다 🙇‍♀️
